fuzzers/libfuzzer-fork-parallel/fuzzer.py

```diff
@@ -0,0 +1,87 @@
+# Copyright 2020 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Integration code for libFuzzer fuzzer."""
+
+import subprocess
+import os
+
+from fuzzers import utils
+
+
+def build():
+    """Build benchmark."""
+    # With LibFuzzer we use -fsanitize=fuzzer-no-link for build CFLAGS and then
+    # /usr/lib/libFuzzer.a as the FUZZER_LIB for the main fuzzing binary. This
+    # allows us to link against a version of LibFuzzer that we specify.
+    cflags = ['-fsanitize=fuzzer-no-link']
+    utils.append_flags('CFLAGS', cflags)
+    utils.append_flags('CXXFLAGS', cflags)
+
+    os.environ['CC'] = 'clang'
+    os.environ['CXX'] = 'clang++'
+    os.environ['FUZZER_LIB'] = '/usr/lib/libFuzzer.a'
+
+    utils.build_benchmark()
+
+
+def fuzz(input_corpus, output_corpus, target_binary):
+    """Run fuzzer. Wrapper that uses the defaults when calling
+    run_fuzzer."""
+    run_fuzzer(input_corpus,
+               output_corpus,
+               target_binary,
+               extra_flags=['-keep_seed=0', '-cross_over_uniform_dist=1'])
+
+
+def run_fuzzer(input_corpus, output_corpus, target_binary, extra_flags=None):
+    """Run fuzzer."""
+    if extra_flags is None:
+        extra_flags = []
+
+    # Seperate out corpus and crash directories as sub-directories of
+    # |output_corpus| to avoid conflicts when corpus directory is reloaded.
+    crashes_dir = os.path.join(output_corpus, 'crashes')
+    output_corpus = os.path.join(output_corpus, 'corpus')
+    os.makedirs(crashes_dir)
+    os.makedirs(output_corpus)
+
+    flags = [
+        '-print_final_stats=1',
+        # `close_fd_mask` to prevent too much logging output from the target.
+        '-close_fd_mask=3',
+        # Run in fork mode to allow ignoring ooms, timeouts, crashes and
+        # continue fuzzing indefinitely.
+        '-fork=8',
+        '-ignore_ooms=1',
+        '-ignore_timeouts=1',
+        '-ignore_crashes=1',
+        '-entropic=0',
```

Comment:
I would prefer to compare against `-entropic=1`, since that is the default libFuzzer config.

If the patch isn't substantially better than entropic mode, I'd rather avoid new complexity in libFuzzer.

Reply:
Yes, I guarantee it is very effective when -entropic=1.
For the test target, I think this strategy is more effective for programs that generate more seeds from the corpus, so you can choose some programs with more lines of code as the test target, such as php-execuate, ffmpeg, libxml2, sqlite3...
For the number of physical cores, I think 8cores, 16core should be set, if possible, 24cores can also be set
I think it takes 8h and above for the test time
I opened the latest and improved version  https://github.com/gtt1995/libfuzzer-cmab-latest

Master Morehouse, when do you start the deployment, these test data may be of great use to me.
Thank you!

Reply:
In my preliminary experiments, the improved version is better than libFuzzer.

Reply:
@morehouse

> I would prefer to compare against `-entropic=1`, since that is the default libFuzzer config.
> 
> If the patch isn't substantially better than entropic mode, I'd rather avoid new complexity in libFuzzer.

Hello,hello,hello